providers/implementations/ciphers/cipher_aes_gcm_hw_ppc.inc

```diff
@@ -23,6 +23,114 @@ static int aes_ppc_gcm_initkey(PROV_GCM_CTX *ctx, const unsigned char *key,
+extern size_t ppc_aes_gcm_decrypt(const unsigned char *in, unsigned char *out, size_t len,
+                         const void *key, unsigned char ivec[16], u64 *Xi);
+
+static inline int IsBigEndian(int i)
+{
+    return ! *((char *)&i);
+}
+
+static inline u32 UTO32(unsigned char *buf)
+{
+    if (IsBigEndian(1))
+        return ((u32) buf[0]) | ((u32) buf[1] << 8) | ((u32) buf[2] << 16) | ((u32) buf[3] << 24);
+    else
+        return ((u32) buf[0] << 24) | ((u32) buf[1] << 16) | ((u32) buf[2] << 8) | ((u32) buf[3]);
+}
+
+static inline u32 add32TOU(unsigned char buf[4], u32 n)
+{
+    u32 r;
+
+    r = UTO32(buf);
+    r += n;
+    if (IsBigEndian(1)) {
+        buf[0] = (unsigned char) r & 0xFF;
+        buf[1] = (unsigned char) (r >> 8) & 0xFF;
+        buf[2] = (unsigned char) (r >> 16) & 0xFF;
+        buf[3] = (unsigned char) (r >> 24) & 0xFF;
+    } else {
+        buf[0] = (unsigned char) (r >> 24) & 0xFF;
+        buf[1] = (unsigned char) (r >> 16) & 0xFF;
+        buf[2] = (unsigned char) (r >> 8) & 0xFF;
+        buf[3] = (unsigned char) r & 0xFF;
+    }
+    return r;
+}
+
+static size_t aes_p10_gcm_crypt(const unsigned char *in, unsigned char *out, size_t len,
+                         const void *key, unsigned char ivec[16], u64 *Xi, int encrypt)
+{
+    int s = 0;
+    int ndone = 0;
+    int ctr_reset = 0;
+    u64 blocks_unused;
+    u64 nb = len / 16;
+    u64 next_ctr = 0;
+    unsigned char ctr_saved[12];
+
+    memcpy(ctr_saved, ivec, 12);
+
+    while (nb) {
+        blocks_unused = (u64) 0xffffffffU + 1 - (u64) UTO32 (ivec + 12);
+        if (nb > blocks_unused) {
+            len = blocks_unused * 16;
+            nb -= blocks_unused;
+            next_ctr = blocks_unused;
+            ctr_reset = 1;
+        } else {
+            len = nb * 16;
+            next_ctr = nb;
+            nb = 0;
+        }
+
+        if (encrypt) {
+            s = ppc_aes_gcm_encrypt(in, out, len, key, ivec, Xi);
+        } else {
+            s = ppc_aes_gcm_decrypt(in, out, len, key, ivec, Xi);
+        }
+
+        /* add counter to ivec */
+        add32TOU(ivec + 12, (u32) next_ctr);
+        if (ctr_reset) {
+            ctr_reset = 0;
+            in += len;
+            out += len;
+        }
+        memcpy(ivec, ctr_saved, 12);
+        ndone += s;
+    }
+
+    return ndone;
+}
+
+size_t ppc_aes_gcm_encrypt_wrap(const unsigned char *in, unsigned char *out, size_t len,
+                         const void *key, unsigned char ivec[16], u64 *Xi)
+{
+    size_t n;
+    int encrypt;
+
+    encrypt = 1;
+    n = aes_p10_gcm_crypt(in, out, len, key, ivec, Xi, encrypt);
+    return n;
+}
+
+size_t ppc_aes_gcm_decrypt_wrap(const unsigned char *in, unsigned char *out, size_t len,
+                         const void *key, unsigned char ivec[16], u64 *Xi)
+{
+    size_t n;
+    int encrypt;
+
+    encrypt = 0;
+    n = aes_p10_gcm_crypt(in, out, len, key, ivec, Xi, encrypt);
```

Comment:
This should be sufficient..
return aes_p10_gcm_crypt(in, out, len, key, ivec, Xi, 0);